### ISSUE TITLE
Can we disable MD009 - Trailing spaces in code blocks?

### ISSUE THREAD

**Issue by Thraka** [DavidAnson/vscode-markdownlint#84]
MD009 - Trailing spaces is triggered in code blocks. Is there a way to disable that?

**Comment by DavidAnson** [comment]
Yes: https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md013---line-length

**Comment by DavidAnson** [comment]
This is how to apply config changes: https://github.com/DavidAnson/vscode-markdownlint#markdownlintconfig

**Comment by Thraka** [comment]
Thank you for the reply. I'm not seeing how to use this rule since MD009 is the one that is triggered, not MD013.

**Comment by DavidAnson** [comment]
Sorry, I read too quickly! The rule you’re hitting does not currently have an exemption for code blocks. Could you please explain why you want this?  In my experience, trailing whitespace is almost never significant in code.

**Comment by Thraka** [comment]
Copy\pasting code and then using select+tab or select+shift+tab will indent or dedent the code which messes with spaces.

**Comment by DavidAnson** [comment]
This rule is about trailing spaces at the end of a line and does not interfere with the indent/dedent process you describe above. If a code block starts without trailing spaces, what you describe above should not add them - and if it does, I think that’s a problem with that functionality.

If you disagree, could you please give a set of steps to perform and some example code to use to reproduce the problem? Thanks!

**Comment by Thraka** [comment]
Yes you're right. That doesn't affect the actual level, it just keeps the spacing level uniform with the current spaces. To make sure this works in a code block under a list, I set `"MD009": { "list_item_empty_lines": true }` but what I've noticed is that if the code already has spaces in there (not an empty line) to match the indentation level of the code.

I think that in my testing and playing around maybe copy/pasting from notepad, visual studio, whatever editor, it put empty spaces that aligned to the current tab level in the code. I'm assuming some editors do that, mine don't so I'm unsure what process lead to that. For example, using `.` for spaces to visualize:

```
somecode()
{
....// four spaces in 
.... 
....// empty line above this one has 4 empty spaces, so that the text aligns to the code indentation.
}
```

Visual Studio seems to delete the spaces on the empty line in the example above. Is that an option someone can change? Perhaps someone has a default to keep the spaces because they like it? Not sure. But I'm sure there are other IDEs (or notepad-like programs even) that would insert spaces automatically. Or perhaps copy/pasting code from a webpage would have spaces in there? Either way, for code, for me, I don't care that much and having to be alerted about cleaning up a copy/paste code block that doesn't visually ruin the code.

**Comment by DavidAnson** [comment]
I think VS Code can be configured to trim extra space automatically. See: https://github.com/Microsoft/vscode/issues/30312

Based on the above, I think MD009 is behaving correctly and don’t think the scenario of pasting from an editor changes the analysis. People (and tools like git) that prefer not to have trailing spaces probably prefer not to have it anywhere. It seems to be syntactically meaningless here, so I would like to leave the rule as is.

If you feel this rule is getting in your way, disabling it is a good solution.

**Comment by nschonni** [comment]
@Thraka if you're looking for a consistent behaviour, you might want to look at https://editorconfig.org/ since you can set the line trim option for all your IDEs (VS, notepad++, VS Code, etc...)

**Comment by Thraka** [comment]
No worries. Thanks for the info! 

@nschonni thanks.

**Comment by Spreadcat** [comment]
> In my experience, trailing whitespace is almost never significant in code.

Found this an stumble over this issue.
In my situation the codeblock contains markdown itself and uses two whitespaces to define a linebreak.

````markdown

## Templates

Here is markdown template.

```markdown
First line.  
Second line.

Paragraph.
```

````

After "First line" there are two whitespaces which require preserving in order to make the template usable.

Thought Just leave this hear without re-opening the issue (though I would love to).
